apis/core/v1alpha2/featuregate_webhook.go

```diff
@@ -0,0 +1,318 @@
+func computeInvalidStabilityPolicyOverridedFeatures(spec FeatureGateSpec, oldObject *FeatureGate) []string {
+	invalidFeatures := sets.String{}
+	for _, featureRef := range spec.Features {
+		permanentlyVoidAllSupportGuaranteesFieldFromOldObj, found := getPermanentlyVoidAllSupportGuaranteesFieldForFeature(oldObject, featureRef.Name)
+		if !found {
+			// Feature doesn't exist in old object and need not be validated
+			continue
+		}
+		if !featureRef.PermanentlyVoidAllSupportGuarantees && permanentlyVoidAllSupportGuaranteesFieldFromOldObj {
+			invalidFeatures.Insert(featureRef.Name)
+		}
+	}
+	return invalidFeatures.List()
+}
+
+// validateFeatureExists checks that a feature can be found in the cluster.
+func (r *FeatureGate) validateFeatureExists(ctx context.Context, c client.Client) field.ErrorList {
+	var allErrors field.ErrorList
+
+	features := &FeatureList{}
+	if err := c.List(ctx, features); err != nil {
+		allErrors = append(allErrors, field.InternalError(field.NewPath("spec").Child("features"), err))
+		return allErrors
+	}
+
+	invalidFeatures := computeFeaturesThatDoNotExist(r.Spec, features)
+
+	if len(invalidFeatures) > 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"), r.Spec.Features,
+			fmt.Sprintf("some features in the FeatureGate spec do not exist in cluster: %v", invalidFeatures)))
+	}
+	return allErrors
+}
+
+// computeFeaturesThatDoNotExist computes and returns features which are not present in the cluster
+func computeFeaturesThatDoNotExist(spec FeatureGateSpec, features *FeatureList) []string {
+	allFeaturesInSpec := sets.String{}
+	for _, feature := range spec.Features {
+		allFeaturesInSpec.Insert(feature.Name)
+	}
+
+	allFeaturesInCluster := sets.String{}
+	for _, feature := range features.Items {
+		allFeaturesInCluster.Insert(feature.Name)
+	}
+	invalidFeatures := allFeaturesInSpec.Difference(allFeaturesInCluster)
+	return invalidFeatures.List()
+}
+
+// validateConflictingFeaturesInFeatureGate validates that the features in spec does not conflict with features gated by
+// existing featuregates.
+func (r *FeatureGate) validateConflictingFeaturesInFeatureGate(ctx context.Context, c client.Client) field.ErrorList {
+	var allErrors field.ErrorList
+
+	featureGates := &FeatureGateList{}
+	if err := c.List(ctx, featureGates); err != nil {
+		allErrors = append(allErrors, field.InternalError(field.NewPath("spec").Child("features"), err))
+		return allErrors
+	}
+
+	conflicts := computeConflictingFeatures(r, featureGates)
+
+	if len(conflicts) > 0 {
+		allErrors = append(allErrors, field.Invalid(field.NewPath("spec").Child("features"),
+			r.Spec.Features, fmt.Sprintf("features %v cannot be gated by multiple featuregates", conflicts)))
+	}
+	return allErrors
+}
+
+// computeConflictingFeatures computes and returns features that conflict with existing feature gates.
+func computeConflictingFeatures(featureGate *FeatureGate, featureGates *FeatureGateList) []string {
```

Comment:
I probably should mention this here even though we talked about it. Even if as a reminder.

I like your suggestion that instead of writing my own validation logic that does the same thing as this webhook, if you could export these computeXXXXX functions, I can use them for my feature plugin. That would be helpful.

Reply:
resolving this as you are going to write your own validation functions for cli plugin